File: lib/java/src/test/java/com/liveprofile/hashring/HashRingTest.java
package com.liveprofile.hashring;

import org.junit.*;
import static org.junit.Assert.*;
import java.math.BigInteger;

public class HashRingTest {
    @Test
    public void testCreateRing() throws HashRingException {
        HashRing ring = new HashRing(1, HashRing.HashFunction.MD5);
    }
    
    @Test
    public void testKnownSlotsOnRing() throws HashRingException {
        HashRing ring = new HashRing(8, HashRing.HashFunction.SHA1);
        assertTrue(ring.addNode("slotA"));
        assertTrue(ring.addNode("slotB"));
        
        assertEquals("slotA", ring.findNode("keyA"));
        assertEquals("slotA", ring.findNode("keyBBBB"));
        assertEquals("slotB", ring.findNode("keyB_"));
    }
    
    @Test
    public void testAddMultipleTimes() throws HashRingException {
        HashRing ring = new HashRing(8, HashRing.HashFunction.SHA1);
        assertTrue(ring.addNode("slotB"));
        assertFalse(ring.addNode("slotB"));
        assertTrue(ring.removeNode("slotB"));
        assertTrue(ring.addNode("slotA"));
    }
    
    @Test
    public void testNextHighestItem() throws HashRingException {
        HashRing ring = new HashRing(8, HashRing.HashFunction.SHA1);
        assertTrue(ring.addNode("slotA"));
        assertTrue(ring.addNode("slotB"));
        assertEquals(2584980261350711786L, ring.findNextHighestItem(2351641940735260693L));
    }
    
    @Test
    public void testAddRemoveNode() throws HashRingException {
        HashRing ring = new HashRing(8, HashRing.HashFunction.SHA1);
        ring.addNode("slotA");
        assertTrue(ring.removeNode("slotA"));
        assertFalse(ring.removeNode("slotA"));
    }
}

File: lib/go/hashring_test.go
package hashring

import (
	"testing"

	"bytes"
)

func TestFindNode(t *testing.T) {
	ring := New(8, SHA1)
	defer ring.Free()

	ring.Add([]byte("slotA"))
	ring.Add([]byte("slotB"))

	var node string
	node = string(ring.FindNode([]byte("keyA")))
	if node != "slotA" {
		t.Fatal(node)
	}
	node = string(ring.FindNode([]byte("keyBBBB")))
	if node != "slotA" {
		t.Fatal(node)
	}
	node = string(ring.FindNode([]byte("keyB_")))
	if node != "slotB" {
		t.Fatal(node)
	}
}

func TestFindNodes(t *testing.T) {
	ring := New(8, SHA1)
	defer ring.Free()

	ring.Add([]byte("slotA"))
	ring.Add([]byte("slotB"))

	var nodes [][]byte

	nodes = ring.FindNodes([]byte("keyA"), 3)
	if len(nodes) != 2 ||
		bytes.Compare(nodes[0], []byte("slotA")) != 0 ||
		bytes.Compare(nodes[1], []byte("slotB")) != 0 {
		t.Fatal(nodes)
	}

	nodes = ring.FindNodes([]byte("keyB*_*_*_"), 3)
	if len(nodes) != 2 ||
		bytes.Compare(nodes[0], []byte("slotB")) != 0 ||
		bytes.Compare(nodes[1], []byte("slotA")) != 0 {
		t.Fatal(nodes)
	}

	ring.Add([]byte("slotC"))

	nodes = ring.FindNodes([]byte("keyA"), 3)
	if len(nodes) != 3 ||
		bytes.Compare(nodes[0], []byte("slotC")) != 0 ||
		bytes.Compare(nodes[1], []byte("slotA")) != 0 ||
		bytes.Compare(nodes[2], []byte("slotB")) != 0 {
		t.Fatal(nodes)
	}

	nodes = ring.FindNodes([]byte("keyB*_*_*_"), 3)
	if len(nodes) != 3 ||
		bytes.Compare(nodes[0], []byte("slotC")) != 0 ||
		bytes.Compare(nodes[1], []byte("slotB")) != 0 ||
		bytes.Compare(nodes[2], []byte("slotA")) != 0 {
		t.Fatal(nodes)
	}

	nodes = ring.FindNodes([]byte("keyA"), 2)
	if len(nodes) != 2 ||
		bytes.Compare(nodes[0], []byte("slotC")) != 0 ||
		bytes.Compare(nodes[1], []byte("slotA")) != 0 {
		t.Fatal(nodes)
	}

	nodes = ring.FindNodes([]byte("keyB*_*_*_"), 2)
	if len(nodes) != 2 ||
		bytes.Compare(nodes[0], []byte("slotC")) != 0 ||
		bytes.Compare(nodes[1], []byte("slotB")) != 0 {
		t.Fatal(nodes)
	}
}
